Add std benchmarks for reference

simonask opened the pull request:
This PR adds benchmarks of std functionality equivalent to `chars::count()`, `chars::to_byte_idx()`, and `chars::from_byte_idx()`. This can be used to measure if this library actually does better than the `std` approach using iterators.

simonask committed:
Add std benchmarks for reference

cessen commented on benches/all.rs:
```diff
@@ -36,6 +36,43 @@ macro_rules! bench_chars {
+        byte_idx -= 1;
+    }
+    // Count the number of chars until the char that begins at `byte_idx`.
+    text.char_indices().take_while(|(i, _)| *i < byte_idx).count()
```
It think this can be simplified to `&text[..byte_idx].chars().count()`, since you already ensured that `byte_idx` is on a char boundary.

cessen commented on benches/all.rs:
```diff
@@ -36,6 +36,43 @@ macro_rules! bench_chars {
     };
 }
 
+fn std_from_byte_idx(text: &str, mut byte_idx: usize) -> usize {
```
I think we can inline this in the test code, by just using a `{ }` code block inside the parenthesis of the `black_box` call.

cessen reviewed (commented):
Over-all looks good to me!  If you could just address the couple of comments, I'll get this merged.

Thanks!

cessen commented on benches/all.rs:
```diff
@@ -27,7 +27,7 @@ macro_rules! bench_chars {
 
             group.bench_function("to_byte_idx", |bench| {
-                let idx = $text.len() - 1;
+                let idx = chars::count($text) - 1;
```
Ah! Thanks for this. Silly mistake. Looks like I made the same mistake in the utf16 and lines benchmarks as well, but I'll handle that separately so we don't clutter this PR.

simonask committed:
Address review feedback

simonask commented:
I've addressed your feedback. I'm seeing some pretty wild differences between `to_byte_idx` and `from_byte_idx` in the `std` versions, which makes me think that there may be smarter ways to do it with `std`, but in both of those cases they're slower than `str_indices` - only `count()` is faster in `std`.

cessen commented:
Looks great!  Thanks for doing the leg work on this.
